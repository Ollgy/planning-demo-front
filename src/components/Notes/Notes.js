import React, { PureComponent } from 'react';
import { connect } from 'react-redux';

import AddButton from '../AddButton';
import Textarea from '../Textarea';
import Note from '../Note';
import Preloader from '../Preloader';
import {
  inputForm,
  clearAnim,
  clearFields,
  deleteNote,
  saveNewNote,
  getNotes
} from '../../modules/Notes';
import { F } from '../../const';
import getformattedTime from '../NewTaskForm/utils/formattedTime';
import styles from './Notes.module.css';

class Notes extends PureComponent {
  constructor(props) {
    super(props);
    this.activeInput = React.createRef();
  }

  async componentDidMount() {
    const { user, getNotes } = this.props;

    await getNotes(user.id);

    this.activeInput.current.focus();
  }

  onInputChange = e => {
    const { inputForm } = this.props;
    const { value, name } = e.target;

    inputForm({ input: name, value });
  };

  saveNewNote = async () => {
    const { fields, saveNewNote, clearAnim, clearFields } = this.props;
    const { text } = fields;

    if (text) {
      await saveNewNote({ text, date: getformattedTime() });
      clearFields();
      setTimeout(clearAnim, 1000);
    }
  };

  deleteNote = async taskId => {
    const { deleteNote } = this.props;

    await deleteNote(taskId);
  };

  render() {
    const { fields, animAppear, notelist, isLoading } = this.props;

    return !isLoading ? (
      <section className={styles.sectionContainer}>
        <div className={styles.panel}>
          <div className={styles.textBlock}>
            <Textarea
              ref={this.activeInput}
              type="text"
              name={F.text}
              value={fields.text}
              onChange={this.onInputChange}
              placeholder="Текст заметки . . ."
              size="big"
            />
          </div>
          <div className={styles.addBlock} data-tooltip="Добавить заметку">
            <AddButton onClick={this.saveNewNote} />
          </div>
        </div>
        {notelist.map(note => (
          <Note
            key={note.id}
            note={note}
            animAppear={animAppear}
            activities={{ deleteNote: this.deleteNote }}
          />
        ))}
      </section>
    ) : (
      <Preloader />
    );
  }
}

const mapStateToProps = state => ({
  fields: state.notes.datalist.fields,
  animAppear: state.notes.datalist.animAppear,
  notelist: state.notes.datalist.notelist,
  isLoading: state.notes.getNotes.isLoading
});

const mapDispatchToProps = dispatch => ({
  saveNewNote: arg => dispatch(saveNewNote(arg)),
  deleteNote: arg => dispatch(deleteNote(arg)),
  getNotes: arg => dispatch(getNotes(arg)),
  inputForm: arg => dispatch(inputForm(arg)),
  clearAnim: () => dispatch(clearAnim()),
  clearFields: () => dispatch(clearFields())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Notes);
