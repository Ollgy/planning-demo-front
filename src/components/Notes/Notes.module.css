.sectionContainer {
  overflow-y: auto;
  height: 100%;
}

.panel {
  display: flex;
  width: 300px;
  justify-content: space-between;
  align-items: center;
  padding: 10px 20px;
  background-color: #efefef;
  margin-bottom: 30px;
}

.addBlock {
  position: relative;
}

.textBlock {
  width: 200px;
}
