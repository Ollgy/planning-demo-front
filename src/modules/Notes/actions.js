import { createAction } from 'redux-actions';
import api from '../../api';

export const inputForm = createAction('INPUT_FORM_NOTE');
export const clearAnim = createAction('CLEAR_ANIM_NOTE');
export const clearFields = createAction('CLEAR_FIELDS_NOTE');

export const saveNewNoteRequest = createAction('SAVE_NEW_NOTE_REQUEST');
export const saveNewNoteSuccess = createAction('SAVE_NEW_NOTE_SUCCESS');
export const saveNewNoteFailure = createAction('SAVE_NEW_NOTE_FAILURE');

export const saveNewNote = noteData => {
  return (dispatch, getState) =>
    new Promise((resolve, reject) => {
      dispatch(saveNewNoteRequest());

      api
        .saveNewNote(noteData)
        .then(notes => {
          dispatch(saveNewNoteSuccess(notes));
          resolve();
        })
        .catch(error => {
          dispatch(saveNewNoteFailure(error));
          reject();
        });
    });
};

export const deleteNoteRequest = createAction('DELETE_NOTE_REQUEST');
export const deleteNoteSuccess = createAction('DELETE_NOTE_SUCCESS');
export const deleteNoteFailure = createAction('DELETE_NOTE_FAILURE');

export const deleteNote = noteId => {
  return (dispatch, getState) =>
    new Promise((resolve, reject) => {
      dispatch(deleteNoteRequest());

      api
        .deleteNote(noteId)
        .then(notes => {
          dispatch(deleteNoteSuccess(notes));
          resolve();
        })
        .catch(error => {
          dispatch(deleteNoteFailure(error));
          reject();
        });
    });
};

export const getNotesRequest = createAction('GET_NOTES_REQUEST');
export const getNotesSuccess = createAction('GET_NOTES_SUCCESS');
export const getNotesFailure = createAction('GET_NOTES_FAILURE');

export const getNotes = userId => {
  return (dispatch, getState) =>
    new Promise((resolve, reject) => {
      dispatch(getNotesRequest());

      api
        .getNotes(userId)
        .then(notes => {
          dispatch(getNotesSuccess(notes));
          resolve();
        })
        .catch(error => {
          dispatch(getNotesFailure(error));
          reject();
        });
    });
};
