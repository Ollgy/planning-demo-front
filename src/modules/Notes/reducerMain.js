import { combineReducers } from 'redux';
import delNote from './reducers/delNote';
import newNote from './reducers/newNote';
import getNotes from './reducers/getNotes';
import datalist from './reducers/datalist';

export default combineReducers({
  datalist,
  newNote,
  delNote,
  getNotes
});
