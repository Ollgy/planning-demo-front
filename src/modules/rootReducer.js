import { combineReducers } from 'redux';
import authFromToken from './Auth';
import login from './Login';
import loginForm from './LoginForm';
import profile from './Profile';
import personalForm from './PersonalForm';
import changePswrdForm from './ChangePswrdForm';
import newTaskForm from './NewTaskForm';
import notes from './Notes';
import dashboard from './Dashboard';
import staff from './Staff';
import cleanBase from './CleanBase';

export default combineReducers({
  authFromToken,
  login,
  loginForm,
  profile,
  personalForm,
  changePswrdForm,
  newTaskForm,
  notes,
  dashboard,
  staff,
  cleanBase
});
