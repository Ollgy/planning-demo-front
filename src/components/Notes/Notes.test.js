import React from 'react';
import Notes from './Notes';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { createStore, compose, applyMiddleware } from 'redux';
import rootReducer from '../../modules/rootReducer';
import thunk from 'redux-thunk';

const store = createStore(rootReducer, {}, compose(applyMiddleware(thunk)));

describe('<Notes/> UI component', () => {
  const shallowExpect = compose(
    expect,
    toJSON,
    shallow
  );

  it('Renders correctly', () => {
    shallowExpect(<Notes store={store} />).toMatchSnapshot();
  });

  const wrapper = shallow(<Notes store={store} />);
  const inner = wrapper.children().first();

  it('contains fields in props', () => {
    expect(inner.props().fields).toBeTruthy();
  });
});
