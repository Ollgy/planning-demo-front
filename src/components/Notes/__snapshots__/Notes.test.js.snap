// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notes/> UI component Renders correctly 1`] = `
<ContextProvider
  value={null}
>
  <Notes
    animAppear={false}
    clearAnim={[Function]}
    clearFields={[Function]}
    deleteNote={[Function]}
    fields={
      Object {
        "text": "",
      }
    }
    getNotes={[Function]}
    inputForm={[Function]}
    isLoading={false}
    notelist={Array []}
    saveNewNote={[Function]}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  />
</ContextProvider>
`;
